File: 백준/Silver/1535. 안녕/안녕.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] health = new int[N];
        int[] joy = new int[N];
        int[][] dp = new int[N + 1][100];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            health[i] = Integer.parseInt(st.nextToken());
        }
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            joy[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= 99; j++) {
                if (health[i - 1] <= j) {
                    dp[i][j] = Math.max(dp[i - 1][j], joy[i - 1] + dp[i - 1][j - health[i - 1]]);
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
            }
        }

        System.out.println(dp[N][99]);
    }
}

File: 백준/Silver/1535. 안녕/안녕.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val n = readLine().toInt()
    val health = IntArray(n)
    val joy = IntArray(n)
    var st = StringTokenizer(readLine())
    for (i in 0 until n) {
        health[i] = st.nextToken().toInt()
    }

    st = StringTokenizer(readLine())
    for (i in 0 until n) {
        joy[i] = st.nextToken().toInt()
    }

    val dp = Array(n + 1) { IntArray(100) { 0 } }
    for (i in 1..n) {
        for (j in 1..99) {
            if (health[i - 1] <= j) {
                dp[i][j] = dp[i - 1][j].coerceAtLeast(joy[i - 1] + dp[i - 1][j - health[i - 1]])
            } else {
                dp[i][j] = dp[i - 1][j];
            }
        }
    }

    print(dp[n][99])
}